### PR TITLE
add offline settings page

### PR DESCRIPTION
**Pull request by lucasxsong**
## Types of changes

What types of changes does this pull request introduce?

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Refactoring (non-breaking change that does not add functionality but makes code cleaner or more efficient)
- [ ] Documentation Update
- [ ] Tests (new or updated tests)
- [ ] Style (changes to code styling)
- [ ] CI (continuous integration changes)
- [ ] Repository Maintenance
- [ ] Other (please describe below)

## Checklist

- [x] I have read the [guidelines for contributing](https://github.com/yeatmanlab/roar-dashboard/blob/main/.github/CONTRIBUTING.md).
- [x] The changes in this PR are as small as they can be. They represent one and only one fix or enhancement.
- [x] Linting checks pass with my changes.
- [x] Any existing unit tests pass with my changes.
- [x] Any existing end-to-end tests pass with my changes.
- [ ] I have added tests that prove my fix is effective or that my feature works.
- [x] If this PR fixes an existing issue, I have added a unit or end-to-end test that will detect if this issue reoccurs.
- [x] I have added JSDoc comments as appropriate.
- [x] I have added the necessary documentation to the [roar-docs repository](https://github.com/yeatmanlab/roar-docs).
- [x] I have shared this PR on the roar-pr-reviews channel (if I have access)
- [x] I have linked relevant issues (if any)

## Justification of missing checklist items

Will add tests when more offline functionality is added.

### CODE REVIEW

**Review by ksmontville — CHANGES_REQUESTED** [review]
I only tested briefly, but this is looking really great.

I am not sure that the delete buttons are properly removing tasks from my offline queue.

Also, one important thing that I think we should add is to re-fetch the data after the user saves their settings. Currently the user has to manually refresh the page after clicking "Save settings" in order to see their changes.

I suggest that clicking the "Save Settings" button triggers a re-fetch using tanstackquery.  I implemented something like this in Manage Tasks/Variants, which re-fetches the data after editing and submitting it to Firestore, after the toast pops up.

You declare a queryClient and then use the invalidateQueries method to manually trigger a re-fetch. Modify to suit your specifi queries.

`import { useQuery, useQueryClient } from '@tanstack/vue-query';
const queryClient = useQueryClient();`

  try {
    await authStore.roarfirekit.updateTaskOrVariant(updateData);
    toast.add({ severity: 'success', summary: 'Hoorah!', detail: 'Variant successfully updated.', life: 3000 });

    // Reset the form and re-fetch the data
    resetUpdateVariantForm();
    await queryClient.invalidateQueries(['tasks', registeredTasksOnly]);
    await queryClient.invalidateQueries(['variants', 'all']);
  } catch (error) {
    console.error(error);
  }
};
`

**Review by kellyel — APPROVED** [review]

File: src/components/views/OfflineSettings.vue [hunk 1]
```diff
@@ -0,0 +1,376 @@
...
+      <div class="flex flex-column">
+        <div
+          v-if="selectedOfflineAdministrations.length === 0"
+          class="px-2 py-4 rounded bg-gray-200 flex align-items-center justify-content-center text-gray-400"
+        >
+          No administrations added
+        </div>
+        <div v-else class="flex flex-column gap-2">
+          <div
+            v-for="name in selectedOfflineAdministrations"
```
Comment (ksmontville) [inline src/components/views/OfflineSettings.vue:116]:
```suggestion
            v-for="name in selectedOfflineAdministrations" :key="name"
```

File: src/components/views/OfflineSettings.vue [hunk 2]
```diff
@@ -0,0 +1,376 @@
...
+        <div class="text-xs text-gray-500">
+          <div>
+            Offline mode is currently under development. Features are limited and being rolled out in an experimental
+            basis.
+          </div>
+          <div>Please note that you may experience more bugs by toggling this feature.</div>
+        </div>
+      </div>
+      <div>
+        <PvToggleButton v-model="offlineEnabled" onLabel="On" offLabel="Off" class="p-2 rounded" />
```
Comment (ksmontville) [inline src/components/views/OfflineSettings.vue:23]:
```suggestion
        <PvToggleButton v-model="offlineEnabled" on-label="On" off-label="Off" class="p-2 rounded" />
```

File: src/components/views/OfflineSettings.vue [hunk 3]
```diff
@@ -0,0 +1,376 @@
...
+        <div v-else class="flex flex-column gap-2">
+          <div
+            v-for="name in selectedOfflineTasks"
+            class="flex justify-content-end font-bold gap-2 p-3 bg-gray-200 rounded"
+          >
+            <PvTag> {{ name }}</PvTag>
+            <div>
+              <PvButton
+                class="text-red-900 border-none bg-gray-100 rounded h-2rem text-sm hover:bg-red-900 hover:text-white"
+                :onClick="() => removeOfflineTask(name)"
```
Comment (ksmontville) [inline src/components/views/OfflineSettings.vue:68]:
```suggestion
                @click="() => removeOfflineTask(name)"
```

File: src/components/views/OfflineSettings.vue [hunk 4]
```diff
@@ -0,0 +1,376 @@
...
+            v-model="selectedOfflineAdministration"
+            :options="administrations"
+            option-label="name"
+            class="h-3rem"
+            option-value="name"
+            placeholder="Select an Administration"
+          />
+          <PvButton
+            class="m-0 bg-primary text-white border-none border-round h-3rem text-sm hover:bg-red-900"
+            :onClick="addOfflineAdministration"
```
Comment (ksmontville) [inline src/components/views/OfflineSettings.vue:100]:
```suggestion
            @click="addOfflineAdministration"
```

File: src/components/views/OfflineSettings.vue [hunk 5]
```diff
@@ -0,0 +1,376 @@
...
+        <div v-else class="flex flex-column gap-2">
+          <div
+            v-for="name in selectedOfflineAdministrations"
+            class="flex justify-content-end font-bold gap-2 p-3 bg-gray-200 rounded"
+          >
+            <PvTag> {{ name }}</PvTag>
+            <div>
+              <PvButton
+                class="text-red-900 bg-gray-100 rounded border-none h-2rem text-sm hover:bg-red-900 hover:text-white"
+                :onClick="() => removeOfflineAdministration(name)"
```
Comment (ksmontville) [inline src/components/views/OfflineSettings.vue:123]:
```suggestion
                @click="() => removeOfflineAdministration(name)"
```

File: src/components/views/OfflineSettings.vue [hunk 6]
```diff
@@ -0,0 +1,376 @@
...
+          <div class="text-sm">You have unsaved changes.</div>
+        </div>
+      </PvAlert>
+    </div>
+    <div class="flex align-items-center justify-content-center mt-2">
+      <div v-if="isSubmitting" class="mr-2">
+        <PvButton
+          disabled
+          class="m-0 bg-primary text-white border-none border-round h-2rem text-md hover:bg-red-900"
+          :onClick="saveOfflineSettings"
```
Comment (ksmontville) [inline src/components/views/OfflineSettings.vue:145]:
```suggestion
          @click="saveOfflineSettings"
```

File: src/components/views/OfflineSettings.vue [hunk 7]
```diff
@@ -0,0 +1,376 @@
...
+        >
+          <i v-if="isSubmitting" class="pi pi-spinner pi-spin mr-2" />
+          Save Settings
+        </PvButton>
+      </div>
+      <div v-else>
+        <PvButton
+          class="m-0 bg-primary text-white border-none border-round h-2rem text-md hover:bg-red-900"
+          :onClick="saveOfflineSettings"
```
Comment (ksmontville) [inline src/components/views/OfflineSettings.vue:154]:
```suggestion
          @click="saveOfflineSettings"
```

File: src/components/views/OfflineSettings.vue [hunk 8]
```diff
@@ -0,0 +1,376 @@
...
+  </div>
+  <PvConfirmDialog />
+</template>
+<script setup>
+import { ref, onMounted, computed, watch } from 'vue';
+import { useAuthStore } from '@/store/auth';
+import { storeToRefs } from 'pinia';
+import { useToast } from 'primevue/usetoast';
+import { fetchDocById } from '@/helpers/query/utils';
+import { useQuery, QueryClient } from '@tanstack/vue-query';
```
Comment (ksmontville) [inline src/components/views/OfflineSettings.vue:170]:
```suggestion
import { useQuery, useQueryClient } from '@tanstack/vue-query';
```

File: src/components/views/OfflineSettings.vue [hunk 9]
```diff
@@ -0,0 +1,376 @@
...
+import { taskFetcher } from '@/helpers/query/tasks';
+import { administrationPageFetcher } from '@/helpers/query/administrations';
+
+// +----------------+
+// | Initialization |
+// +----------------+
+const authStore = useAuthStore();
+const toast = useToast();
+const { roarfirekit, uid, administrationQueryKeyIndex, userClaimsQueryKeyIndex } = storeToRefs(authStore);
+const queryClient = new QueryClient();
```
Comment (ksmontville) [inline src/components/views/OfflineSettings.vue:181]:
```suggestion
const queryClient = useQueryClient()
```

File: src/components/views/OfflineSettings.vue [hunk 10]
```diff
@@ -0,0 +1,376 @@
...
+      toast.add({
+        severity: 'error',
+        summary: 'Unexpected Error',
+        detail: 'An unexpected error has occurred.',
+        life: 3000,
+      });
+    });
+  // invalidate tanstack queries
+
+  queryClient.invalidateQueries(['userData', uid]);
```
Comment (ksmontville) [inline src/components/views/OfflineSettings.vue:343]:
```suggestion
await queryClient.invalidateQueries(['userData', uid]);
```
Either `await` this or move it into the .then() block above (don't think it matters which one).

**Review by ksmontville — CHANGES_REQUESTED** [review]
Looking good!

**Review by ksmontville — APPROVED** [review]

**Pull request merged**